### ISSUE TITLE
Packages could not be mapped to the NPM lockfile.

### ISSUE THREAD

**Issue by dpotera** [nrwl/nx#28832]
### Current Behavior

I have npm [package](https://github.com/dpotera/npm-tests) (with some sub-dependencies), which I'm using in nestjs app.
During build of app creation of lock file fails.
After removal of this package usage, problem is not happening.

### Expected Behavior

I should be able to use npm package and lock file should be created successfully.

### GitHub Repo

https://github.com/dpotera/nx-npm-example

### Steps to Reproduce

1. clone repository
2. `npm install`
3. `npx nx build example-api` - prints error

### Nx Report

 NX   Report complete - copy this into the issue template

Node           : 22.0.0
OS             : darwin-arm64
Native Target  : aarch64-macos
npm            : 10.5.1

nx (global)        : 20.0.0
nx                 : 20.0.10
@nx/js             : 20.0.10
@nx/jest           : 20.0.10
@nx/eslint         : 20.0.10
@nx/workspace      : 20.0.10
@nx/devkit         : 20.0.10
@nx/eslint-plugin  : 20.0.10
@nx/nest           : 20.0.10
@nx/node           : 20.0.10
@nx/web            : 20.0.10
@nx/webpack        : 20.0.10
typescript         : 5.5.4
---------------------------------------
Registered Plugins:
@nx/webpack/plugin
@nx/eslint/plugin
@nx/jest/plugin

### Failure Logs

> nx run example-api:build  [existing outputs match the cache, left as is]

> webpack-cli build --node-env=production

 NX   An error occured while creating pruned lockfile

Please open an issue at `https://github.com/nrwl/nx/issues/new?template=1-bug.yml` and provide a reproduction.
To prevent the build from breaking we are returning the root lock file.
If you run `npm install --package-lock-only` in your output folder it will regenerate the correct pruned lockfile.

Original error: Following packages could not be mapped to the NPM lockfile:
- npm:uuid@9.0.1
- npm:@aws-crypto/supports-web-crypto@5.2.0
- npm:@aws-crypto/util@5.2.0
- npm:@aws-sdk/types@3.686.0
- npm:@smithy/types@3.6.0
- npm:@aws-crypto/sha256-browser@5.2.0
- npm:@aws-crypto/sha256-js@5.2.0
- npm:@aws-sdk/client-sso-oidc@3.686.0
- npm:@aws-sdk/client-sts@3.686.0
- npm:@aws-sdk/core@3.686.0
- npm:@smithy/middleware-serde@3.0.8
- npm:@smithy/protocol-http@4.1.5
- npm:@smithy/util-body-length-browser@3.0.0
- npm:@smithy/util-middleware@3.0.8
- npm:@smithy/util-stream@3.2.1
- npm:@smithy/fetch-http-handler@4.0.0
- npm:@smithy/querystring-builder@3.0.8
- npm:@smithy/util-uri-escape@3.0.0
- npm:@smithy/util-base64@3.0.0
- npm:@smithy/util-buffer-from@3.0.0
- npm:@smithy/is-array-buffer@3.0.0
- npm:@smithy/util-utf8@3.0.0
- npm:@smithy/node-http-handler@3.2.5
- npm:@smithy/abort-controller@3.1.6
- npm:@smithy/util-hex-encoding@3.0.0
- npm:@smithy/node-config-provider@3.1.9
- npm:@smithy/property-provider@3.1.8
- npm:@smithy/shared-ini-file-loader@3.1.9
- npm:@smithy/signature-v4@4.2.1
- npm:@smithy/smithy-client@3.4.2
- npm:@smithy/middleware-endpoint@3.2.1
- npm:@smithy/url-parser@3.0.8
- npm:@smithy/querystring-parser@3.0.8
- npm:@smithy/middleware-stack@3.0.8
- npm:fast-xml-parser@4.4.1
- npm:@aws-sdk/credential-provider-node@3.686.0
- npm:@aws-sdk/credential-provider-env@3.686.0
- npm:@aws-sdk/credential-provider-ini@3.686.0
- npm:@aws-sdk/credential-provider-process@3.686.0
- npm:@aws-sdk/credential-provider-sso@3.686.0
- npm:@aws-sdk/client-sso@3.686.0
- npm:@aws-sdk/middleware-host-header@3.686.0
- npm:@aws-sdk/middleware-logger@3.686.0
- npm:@aws-sdk/middleware-recursion-detection@3.686.0
- npm:@aws-sdk/middleware-user-agent@3.686.0
- npm:@aws-sdk/util-endpoints@3.686.0
- npm:@smithy/util-endpoints@2.1.4
- npm:@aws-sdk/region-config-resolver@3.686.0
- npm:@smithy/util-config-provider@3.0.0
- npm:@aws-sdk/util-user-agent-browser@3.686.0
- npm:@aws-sdk/util-user-agent-node@3.686.0
- npm:@smithy/config-resolver@3.0.10
- npm:@smithy/hash-node@3.0.8
- npm:@smithy/invalid-dependency@3.0.8
- npm:@smithy/middleware-content-length@3.0.10
- npm:@smithy/middleware-retry@3.0.25
- npm:@smithy/service-error-classification@3.0.8
- npm:@smithy/util-retry@3.0.8
- npm:@smithy/util-body-length-node@3.0.0
- npm:@smithy/util-defaults-mode-browser@3.0.25
- npm:@smithy/util-defaults-mode-node@3.0.25
- npm:@smithy/credential-provider-imds@3.2.5
- npm:@aws-sdk/token-providers@3.686.0
- npm:@aws-sdk/credential-provider-web-identity@3.686.0

Error: Following packages could not be mapped to the NPM lockfile:
- npm:uuid@9.0.1
- npm:@aws-crypto/supports-web-crypto@5.2.0
- npm:@aws-crypto/util@5.2.0
- npm:@aws-sdk/types@3.686.0
- npm:@smithy/types@3.6.0
- npm:@aws-crypto/sha256-browser@5.2.0
- npm:@aws-crypto/sha256-js@5.2.0
- npm:@aws-sdk/client-sso-oidc@3.686.0
- npm:@aws-sdk/client-sts@3.686.0
- npm:@aws-sdk/core@3.686.0
- npm:@smithy/middleware-serde@3.0.8
- npm:@smithy/protocol-http@4.1.5
- npm:@smithy/util-body-length-browser@3.0.0
- npm:@smithy/util-middleware@3.0.8
- npm:@smithy/util-stream@3.2.1
- npm:@smithy/fetch-http-handler@4.0.0
- npm:@smithy/querystring-builder@3.0.8
- npm:@smithy/util-uri-escape@3.0.0
- npm:@smithy/util-base64@3.0.0
- npm:@smithy/util-buffer-from@3.0.0
- npm:@smithy/is-array-buffer@3.0.0
- npm:@smithy/util-utf8@3.0.0
- npm:@smithy/node-http-handler@3.2.5
- npm:@smithy/abort-controller@3.1.6
- npm:@smithy/util-hex-encoding@3.0.0
- npm:@smithy/node-config-provider@3.1.9
- npm:@smithy/property-provider@3.1.8
- npm:@smithy/shared-ini-file-loader@3.1.9
- npm:@smithy/signature-v4@4.2.1
- npm:@smithy/smithy-client@3.4.2
- npm:@smithy/middleware-endpoint@3.2.1
- npm:@smithy/url-parser@3.0.8
- npm:@smithy/querystring-parser@3.0.8
- npm:@smithy/middleware-stack@3.0.8
- npm:fast-xml-parser@4.4.1
- npm:@aws-sdk/credential-provider-node@3.686.0
- npm:@aws-sdk/credential-provider-env@3.686.0
- npm:@aws-sdk/credential-provider-ini@3.686.0
- npm:@aws-sdk/credential-provider-process@3.686.0
- npm:@aws-sdk/credential-provider-sso@3.686.0
- npm:@aws-sdk/client-sso@3.686.0
- npm:@aws-sdk/middleware-host-header@3.686.0
- npm:@aws-sdk/middleware-logger@3.686.0
- npm:@aws-sdk/middleware-recursion-detection@3.686.0
- npm:@aws-sdk/middleware-user-agent@3.686.0
- npm:@aws-sdk/util-endpoints@3.686.0
- npm:@smithy/util-endpoints@2.1.4
- npm:@aws-sdk/region-config-resolver@3.686.0
- npm:@smithy/util-config-provider@3.0.0
- npm:@aws-sdk/util-user-agent-browser@3.686.0
- npm:@aws-sdk/util-user-agent-node@3.686.0
- npm:@smithy/config-resolver@3.0.10
- npm:@smithy/hash-node@3.0.8
- npm:@smithy/invalid-dependency@3.0.8
- npm:@smithy/middleware-content-length@3.0.10
- npm:@smithy/middleware-retry@3.0.25
- npm:@smithy/service-error-classification@3.0.8
- npm:@smithy/util-retry@3.0.8
- npm:@smithy/util-body-length-node@3.0.0
- npm:@smithy/util-defaults-mode-browser@3.0.25
- npm:@smithy/util-defaults-mode-node@3.0.25
- npm:@smithy/credential-provider-imds@3.2.5
- npm:@aws-sdk/token-providers@3.686.0
- npm:@aws-sdk/credential-provider-web-identity@3.686.0
    at nestMappedPackages (/Users/dominik/code/nx-npm-workspace/node_modules/nx/src/plugins/js/lock-file/npm-parser.js:367:15)
    at nestMappedPackages (/Users/dominik/code/nx-npm-workspace/node_modules/nx/src/plugins/js/lock-file/npm-parser.js:373:9)
    at nestMappedPackages (/Users/dominik/code/nx-npm-workspace/node_modules/nx/src/plugins/js/lock-file/npm-parser.js:373:9)
    at mapSnapshots (/Users/dominik/code/nx-npm-workspace/node_modules/nx/src/plugins/js/lock-file/npm-parser.js:309:9)
    at stringifyNpmLockfile (/Users/dominik/code/nx-npm-workspace/node_modules/nx/src/plugins/js/lock-file/npm-parser.js:239:28)
    at createLockFile (/Users/dominik/code/nx-npm-workspace/node_modules/nx/src/plugins/js/lock-file/lock-file.js:172:58)
    at /Users/dominik/code/nx-npm-workspace/node_modules/@nx/webpack/src/plugins/generate-package-json-plugin.js:40:138
    at fn (/Users/dominik/code/nx-npm-workspace/node_modules/webpack/lib/Compilation.js:497:10)
    at _next0 (eval at create (/Users/dominik/code/nx-npm-workspace/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:11:1)
    at eval (eval at create (/Users/dominik/code/nx-npm-workspace/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:24:1)

chunk (runtime: main) main.js (main) 2.84 KiB [entry] [rendered]
webpack compiled successfully (40f57bfe94298383)

### Package Manager Version

npm version 10.5.1

### Operating System

- [x] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

**Comment by GrayedFox** [comment]
We have the same problem except we're referencing and installing a local package i.e. we have `local_node_modules/our-package/` inside the project root which is referenced from the root package.json like so:

```json
{
  "our-package": "file:local_node_modules/our-package"
}
```

And our package has some dependencies too, including `express`, so we see a really similar output before the same "could not be mapped" error.

**Comment by GrayedFox** [comment]
Okay we worked around it on our end by ensuring that all `aws-sdk` modules have the exact same matching version number inside our root `package.json` file and converting our local NPM module to an internal ESM module, examples in below JSON are placeholders:

```json
"dependencies": {
        "@aws-sdk/client-s3": "^3.693.0",
        "@aws-sdk/example: "^3.693.0",
        "@aws-sdk/example-2: "^3.693.0",
}
```

Previously our local NPM module specified `@aws-sdk/client-s3" as a peer dependency and at a lower version than what our core root `package.json` file listed as a cored dependency.

I was barking up the wrong tree before, I assumed it was something to do with how NX resolved local NPM modules. The underlying issue was that our local module specified a different version of the same AWS SDK library as a peer dep. NPM will normally error (earlier versions used to warn) if a require or import statement doesn't resolve to anything but for standard dependencies it resolves the import by checking the entire tree. If a matching major version is found, technically the import or require statement will resolve to something, not sure if NPM warns or complains in this case.

For peer dependencies the check is different, NPM treats this field as an exact match (minor and patch versions as well), and it's up to the developer to ensure that matching peer dependency is available, since it's not installed automatically.

I believe what was happening for us is that when NX attempts to generate a `package.lock` file it checks for unresolved parents via a [strict parent lookup](https://github.com/nrwl/nx/blob/master/packages/nx/src/plugins/js/lock-file/npm-parser.ts#L587) based on its own internally generated dependency graph. After NX does some form of tree-shaking to generate the reduced `package.json` for that package, NX sees that there's no matching parent for the peer dependency and any underlying dependencies of that peer dependency, but perhaps @leosvelperez or someone from the NX side could confirm this? Will amend this comment if that's the case.

In short, at least in our case, I don't think NX was bugging out.

@dpotera your examples don't include peer deps, so there may still be a bug here. I noticed that in your core application the common, core, and rxjs deps are lower than the ones included in your `npm-tests` package:

![Image](https://github.com/user-attachments/assets/df25e515-ef2b-4b3f-9331-0bfa87b6243e)

You could try updating your core deps to match the ones in your other package as a workaround?

```
    "@nestjs/common": "^10.4.5",
    "@nestjs/core": "^10.4.5",
    "rxjs": "^7.8.1"
```

**Comment by dpotera** [comment]
Thanks @GrayedFox for your feedback on this issue.

I've removed nestjs and rxjs dependencies from my example package [`npm-tests`](https://github.com/dpotera/npm-tests/blob/541fbc4df5239fbc5abf5418d223987253ec6fea/package.json#L31) dependencies, and simplified them to only include:
```json
{
  "dependencies": {
    "@aws-sdk/client-dynamodb": "3.438.0",
    "@aws-sdk/lib-dynamodb": "3.438.0",
    "snowflake-sdk": "^1.13.1",
    "uuid": "^9.0.1"
  }
}
```
But even after that, the problem still occurs. The packages that failed to map are related to aws clients and their dependencies.

I think the source of the problem is that we include directly aws clients, like `@aws-sdk/client-dynamodb` in version **3.438.0** and `snowflake-sdk` package also includes one of aws clients as a dependency in version **3.388.0**.

`package-lock.json` [link](https://github.com/dpotera/npm-tests/blob/541fbc4df5239fbc5abf5418d223987253ec6fea/package-lock.json#L5646):
```json
{
  "node_modules/snowflake-sdk": {
    "version": "1.14.0",
    "resolved": "https://registry.npmjs.org/snowflake-sdk/-/snowflake-sdk-1.14.0.tgz",
    "integrity": "sha512-dr+Qg+cV1tDayPrYkOg4/t5ARzco6hEIm+ylvxAWg5RWTxbRcuCFVxBx0p1vrf0wFS2igz06qCI4l7ZoM6dsTw==",
    "dependencies": {
      "@aws-sdk/client-s3": "^3.388.0",
      "@aws-sdk/node-http-handler": "^3.374.0",
      "@azure/storage-blob": "12.18.x",
      "...more": "dependencies..."
    }
  }
}
```
It's not possible to align all sub-dependencies to the same version, because they are not controlled by us and are using caret versions. 
Different versions of the same packages in different dependencies are common case in npm environment, and should be supported during dynamic lockfile creation.

**Comment by GrayedFox** [comment]
Yeah we ended up working around this by adding a local_node_modules folder of the dependency that was had mismatched `@aws-sdk/client` versions (luckily we just had the one) so that we could gain control of the `package.json` file. Certainly not a long term solution but allowed us to get around the prune lock file error from NX.

**Comment by sergio-milu** [comment]
I'm seeing this same issue after updating NX from 18.2 to 19.8

```
Error: Following packages could not be mapped to the NPM lockfile:
- npm:uuid@9.0.1
- npm:@aws-crypto/sha256-browser@5.2.0
- npm:@aws-crypto/sha256-js@5.2.0
- npm:@aws-crypto/util@5.2.0
- npm:@aws-sdk/types@3.734.0
- npm:@smithy/types@4.1.0
- npm:@aws-sdk/types@3.654.0
- npm:@smithy/types@3.4.2
- npm:@aws-sdk/types@3.686.0
- npm:@smithy/types@3.6.0
- npm:@aws-crypto/supports-web-crypto@5.2.0
- npm:fast-xml-parser@4.4.1
- npm:@aws-sdk/client-sso-oidc@3.687.0
- npm:@aws-sdk/client-sts@3.687.0
- npm:@aws-sdk/core@3.686.0
- npm:@smithy/core@2.5.1
- npm:@smithy/middleware-serde@3.0.8
- npm:@smithy/protocol-http@4.1.5
- npm:@smithy/util-body-length-browser@3.0.0
- npm:@smithy/util-middleware@3.0.8
- npm:@smithy/util-stream@3.2.1
- npm:@smithy/fetch-http-handler@4.0.0
- npm:@smithy/querystring-builder@3.0.8
- npm:@smithy/util-uri-escape@3.0.0
- npm:@smithy/util-base64@3.0.0
- npm:@smithy/util-buffer-from@3.0.0
- npm:@smithy/is-array-buffer@3.0.0
- npm:@smithy/util-utf8@3.0.0
- npm:@smithy/node-http-handler@3.2.5
- npm:@smithy/abort-controller@3.1.6
- npm:@smithy/util-hex-encoding@3.0.0
- npm:@smithy/node-config-provider@3.1.9
- npm:@smithy/property-provider@3.1.8
- npm:@smithy/shared-ini-file-loader@3.1.9
- npm:@smithy/signature-v4@4.2.1
- npm:@smithy/smithy-client@3.4.2
- npm:@smithy/middleware-endpoint@3.2.1
- npm:@smithy/url-parser@3.0.8
- npm:@smithy/querystring-parser@3.0.8
- npm:@smithy/middleware-stack@3.0.8
- npm:@aws-sdk/credential-provider-node@3.687.0
- npm:@aws-sdk/credential-provider-env@3.686.0
- npm:@aws-sdk/credential-provider-http@3.686.0
- npm:@aws-sdk/credential-provider-ini@3.687.0
- npm:@aws-sdk/credential-provider-process@3.686.0
- npm:@aws-sdk/credential-provider-sso@3.687.0
- npm:@aws-sdk/client-sso@3.687.0
- npm:@aws-sdk/middleware-host-header@3.686.0
- npm:@aws-sdk/middleware-logger@3.686.0
- npm:@aws-sdk/middleware-recursion-detection@3.686.0
- npm:@aws-sdk/middleware-user-agent@3.687.0
- npm:@aws-sdk/util-endpoints@3.686.0
- npm:@smithy/util-endpoints@2.1.4
- npm:@aws-sdk/region-config-resolver@3.686.0
- npm:@smithy/util-config-provider@3.0.0
- npm:@aws-sdk/util-user-agent-browser@3.686.0
- npm:@aws-sdk/util-user-agent-node@3.687.0
- npm:@smithy/config-resolver@3.0.10
- npm:@smithy/hash-node@3.0.8
- npm:@smithy/invalid-dependency@3.0.8
- npm:@smithy/middleware-content-length@3.0.10
- npm:@smithy/middleware-retry@3.0.25
- npm:@smithy/service-error-classification@3.0.8
- npm:@smithy/util-retry@3.0.8
- npm:@smithy/util-body-length-node@3.0.0
- npm:@smithy/util-defaults-mode-browser@3.0.25
- npm:@smithy/util-defaults-mode-node@3.0.25
- npm:@smithy/credential-provider-imds@3.2.5
- npm:@aws-sdk/token-providers@3.686.0
- npm:@aws-sdk/credential-provider-web-identity@3.686.0
    at nestMappedPackages (/Users/sergiofernandez/Repositories/milu-web/node_modules/nx/src/plugins/js/lock-file/npm-parser.js:367:15)
```

**Comment by sergio-milu** [comment]
I think I managed to solved this issue aligning all aws skd packages to the same version (3.777.0) in my case
